url85940d3d8e_ccp999_Expr4j_tgz-pJ8-org_boris_expr_util_ExprEvaluatorJ8/src/org/boris/expr/function/excel/ADDRESS.java

```diff
@@ -0,0 +1,80 @@
+package org.boris.expr.function.excel;
+
+import org.boris.expr.Expr;
+import org.boris.expr.ExprException;
+import org.boris.expr.ExprString;
+import org.boris.expr.IEvaluationContext;
+import org.boris.expr.engine.GridReference;
+import org.boris.expr.function.AbstractFunction;
+
+public class ADDRESS extends AbstractFunction
+{
+    public Expr evaluate(IEvaluationContext context, Expr[] args)
+            throws ExprException {
+        if (args.length < 2 || args.length > 5)
+            throw new ExprException("Invalid arguments to function ADDRESS");
+
+        int row = asInteger(context, args[0], true);
+        int col = asInteger(context, args[1], true);
+        int abs = 1;
+        if (args.length >= 3) {
+            abs = asInteger(context, args[2], true);
+        }
+        boolean a1 = true;
+        if (args.length >= 4) {
+            a1 = asBoolean(context, args[3], true);
+        }
+        String sheetText = null;
+        boolean needsQuotes = false;
+        if (args.length >= 5) {
+            sheetText = asString(context, args[4], true);
+            needsQuotes = needsQuotes(sheetText);
+        }
+
+        StringBuilder sb = new StringBuilder();
+        if (sheetText != null) {
+            if (needsQuotes)
+                sb.append('\'');
+            sb.append(sheetText);
+            if (needsQuotes)
+                sb.append('\'');
+            sb.append('!');
+        }
+        if (a1) {
+            if (abs == 1 || abs == 3)
+                sb.append('$');
+            sb.append(GridReference.toColumnName(col));
+            if (abs == 1 || abs == 2)
+                sb.append('$');
+            sb.append(row);
+        } else {
+            sb.append('R');
+            if (abs == 3 || abs == 4) {
+                sb.append('[');
+                sb.append(row);
+                sb.append(']');
+            } else {
+                sb.append(row);
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 52:58, 60:66.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.